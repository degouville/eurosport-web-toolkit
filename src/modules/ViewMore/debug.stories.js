// TODO: This file is here for the moment because we will have to test this behaviour we the new design

import React, { useState, useEffect, useRef } from 'react';
import { storiesOf } from '@storybook/react';
import styled, { css } from 'react-emotion';

// eslint-disable-next-line import/no-extraneous-dependencies
import { Global } from '@emotion/core';
import { object, text, number, select } from '@storybook/addon-knobs';
import { random } from 'lodash';
import { ViewMore, Button } from 'src';
import { white } from '../../colors';
import SimpleTabs from '../SimpleTabs';

const viewmore = storiesOf('Modules|ViewMore', module);

const labels = {
  showLess: 'View less',
  showMore: 'View more',
};

const Wrapper = styled.div`
  color: ${white};
`;

const StyledElement = styled.div`
  margin-bottom: 15px;
  height: ${({ height: h }) => h}px;
  transition: height 1s;
`;

const useInterval = (callback, delay) => {
  const savedCallback = useRef();

  useEffect(() => {
    savedCallback.current = callback;
  }, [callback]);

  // eslint-disable-next-line consistent-return
  useEffect(() => {
    // eslint-disable-next-line consistent-return
    const tick = () => savedCallback.current();
    if (delay) {
      const id = setInterval(tick, delay);
      return () => clearInterval(id);
    }
  }, [delay]);
};

// eslint-disable-next-line react/prop-types
const Element = ({ id }) => {
  const min = 16;
  const [height, setHeight] = useState(min);
  const randomHeight = random(min, min * (id + 1));

  useInterval(() => setHeight(randomHeight), 2000);

  return <StyledElement height={height}>Match</StyledElement>;
};

const createChildren = (x, i) => <Element key={i} id={i} />;
const createChildrens = n =>
  Array(n)
    .fill()
    .map(createChildren);

const childrens = [createChildrens(8), createChildrens(4)];

const mockData = [
  {
    label: '1st Round',
    highlighted: true,
    href: '#1',
  },
  {
    label: '2nd Round',
    highlighted: true,
    href: '#2',
  },
];

class TabExample extends React.Component {
  state = {
    currentTab: 1,
    filter: null,
  };

  componentWillReceiveProps(nextProps) {
    // eslint-disable-next-line react/prop-types
    const { initialTab } = nextProps;
    this.setState({ currentTab: initialTab });
  }

  changeTab(href, tabIndex) {
    this.setState({ currentTab: tabIndex });
  }

  filteredChildrens() {
    const { currentTab, filter } = this.state;
    const hasIndexAbove = (x, i) => i <= filter;
    return filter ? childrens[currentTab].filter(hasIndexAbove) : childrens[currentTab];
  }

  updateFilter() {
    const { currentTab } = this.state;
    const childrensCount = childrens[currentTab].length;
    const filter = random(1, childrensCount);
    this.setState({ filter });
  }

  render() {
    const { currentTab } = this.state;
    return (
      <>
        <Global
          styles={css`
            p {
              color: red;
            }
            ,
            .tabItem {
              display: none;
              color: white;
              padding-top: 40px;
              font-size: 30px;
            }
          `}
        />

        <Button onClick={() => this.updateFilter()}>UPDATE MATCHES</Button>

        <SimpleTabs
          initialTabIndex={currentTab}
          /* eslint-disable-next-line react/jsx-no-bind */
          onItemSelected={this.changeTab.bind(this)}
          tabs={object('tabs', mockData)}
        />
        <p>{`There is ${this.filteredChildrens().length} loaded matches.`}</p>

        <Wrapper>
          {/* childrens[currentTab] */}
          <ViewMore
            showMoreText={text('showMoreText', labels.showMore)}
            showLessText={text('showLessText', labels.showLess)}
            showLessItemCount={number('showLessItemCount', 1)}
          >
            {this.filteredChildrens()}
          </ViewMore>
        </Wrapper>
      </>
    );
  }
}

viewmore.add('debug viewmore', () => (
  <TabExample initialTab={select('initialTabIndex', mockData.map((el, i) => i), 2)} />
));
