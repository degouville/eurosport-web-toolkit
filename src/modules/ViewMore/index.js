import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';
import { rgba } from 'polished';
import get from 'lodash/get';
import throttle from 'lodash/throttle';
import { coreLightMinus1, coreLightBase, coreDarkPlus1 } from '../../colors';
import { fontAlphaHeadlineBold } from '../../typography';

export const StyledButton = styled.span`
  color: ${coreLightBase};
  ${fontAlphaHeadlineBold};
  border: 1px solid ${rgba(coreLightMinus1, 0.35)};
  background: ${rgba(coreDarkPlus1, 0.2)};
  border-radius: 3px;
  text-transform: uppercase;
  font-size: 11px;
  padding: 6px 14px;
  line-height: 11px;
  &:hover {
    cursor: pointer;
  }
`;

const StyledList = styled.ul`
  overflow: hidden;
  ${props =>
    css`
      max-height: ${props.blockHeight !== 0 ? `${props.blockHeight}px;` : 'auto;'}
      transition: ${props.animate ? 'max-height 400ms ease;' : 'none;'}
    `}
`;

export default class ViewMore extends React.Component {
  listRef = React.createRef();

  state = {
    animate: true,
  };

  handleWindowResize = throttle(() => this.setState({ animate: false }), 250);

  componentDidMount() {
    // It is better to set the default value of expanded after the initial mount.
    // Otherwise the first render will not be accurate as the listRef might not be set yet.
    this.setState({ expanded: false });
    window.addEventListener('resize', this.handleWindowResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleWindowResize);
  }

  computeBlockHeight() {
    const { showLessItemCount } = this.props;
    const { expanded } = this.state;
    const listElement = get(this, 'listRef.current', {});
    const listChildren = get(listElement, 'children', []);

    let height = 0;
    if (listChildren.length > 0)
      height =
        expanded || showLessItemCount >= listChildren.length
          ? listElement.scrollHeight
          : listChildren[showLessItemCount].offsetTop - listElement.offsetTop;

    return height;
  }

  handleClick() {
    this.setState(state => ({
      expanded: !state.expanded,
      animate: true,
    }));
  }

  render() {
    const { children, showLessText, showMoreText } = this.props;
    const { expanded, animate } = this.state;
    if (!children || !children.length) return null;
    return (
      <>
        <StyledList
          isExpanded={expanded}
          innerRef={this.listRef}
          blockHeight={this.computeBlockHeight()}
          animate={animate}
        >
          {children}
        </StyledList>
        {children.length > 1 && (
          <StyledButton onClick={() => this.handleClick()}>{expanded ? showLessText : showMoreText}</StyledButton>
        )}
      </>
    );
  }
}

ViewMore.defaultProps = {
  showLessItemCount: 1,
  showLessText: 'show less',
  showMoreText: 'show more',
};

ViewMore.propTypes = {
  children: PropTypes.arrayOf(PropTypes.node).isRequired,
  showLessText: PropTypes.string,
  showMoreText: PropTypes.string,
  showLessItemCount: PropTypes.number,
};
