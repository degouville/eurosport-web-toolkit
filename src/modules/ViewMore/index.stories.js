import React from 'react';
import { storiesOf } from '@storybook/react';
import styled from 'react-emotion';
import { object, text } from '@storybook/addon-knobs';
import { ViewMore } from '../..';
import { coreLightMinus1 } from '../../colors';

const viewmore = storiesOf('Modules|ViewMore', module);

const labels = {
  showLess: 'View less',
  showMore: 'View more',
};

const Wrapper = styled.div`
  color: ${coreLightMinus1};
`;

const Element = styled.div`
  margin-bottom: 15px;
`;

const children = [];
for (let i = 0; i < 10; i += 1) {
  children.push(<Element key={i}>match</Element>);
}

viewmore.add('viewmore', () => (
  <Wrapper>
    <ViewMore showMoreText={text('showMoreText', labels.showMore)} showLessText={text('showLessText', labels.showLess)}>
      {object('children', children)}
    </ViewMore>
  </Wrapper>
));
