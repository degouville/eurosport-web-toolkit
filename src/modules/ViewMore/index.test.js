import React from 'react';
import { mount } from 'enzyme';
import ViewMore, { StyledViewMoreButton } from '.';

const labels = {
  showLess: 'View less',
  showMore: 'View more',
};

const children = [];
for (let i = 0; i < 10; i += 1) {
  children.push(<div key={i}>match</div>);
}

describe('ViewMore', () => {
  it('renders its content', () => {
    expect(
      mount(
        <ViewMore showLessText={labels.showLess} showMoreText={labels.showMore}>
          {children}
        </ViewMore>
      )
    ).toMatchSnapshot();
  });
  describe('items behavior', () => {
    it('should display the button with the right label', () => {
      const wrapper = mount(
        <ViewMore showLessText={labels.showLess} showMoreText={labels.showMore}>
          {children}
        </ViewMore>
      );
      wrapper.setState({ expanded: true });
      expect(wrapper.find(StyledViewMoreButton).text()).toContain(labels.showLess);
      wrapper.setState({ expanded: false });
      expect(wrapper.find(StyledViewMoreButton).text()).toContain(labels.showMore);
    });

    it('should update the state when the button is clicked', () => {
      const wrapper = mount(
        <ViewMore showLessText={labels.showLess} showMoreText={labels.showMore}>
          {children}
        </ViewMore>
      );
      expect(wrapper.state().expanded).toEqual(false);
      wrapper.find(StyledViewMoreButton).simulate('click');
      expect(wrapper.state().expanded).toEqual(true);
      wrapper.find(StyledViewMoreButton).simulate('click');
      expect(wrapper.state().expanded).toEqual(false);
    });
  });
});
