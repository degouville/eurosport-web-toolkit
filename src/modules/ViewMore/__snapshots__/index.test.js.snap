// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewMore renders its content 1`] = `
.emotion-0 {
  overflow: hidden;
  height: auto;
  -webkit-transition: height 400ms ease;
  transition: height 400ms ease;
}

.emotion-2 {
  color: #f8f8f9;
  font-family: 'ESP AlphaHeadline Tab',-apple-system,-system-ui,system-ui,BlinkMacSystemFont,'Segoe UI',Helvetica,Arial,sans-serif;
  -webkit-font-smoothing: auto;
  -moz-osx-font-smoothing: auto;
  font-smooth: auto;
  border: 1px solid rgba(255,255,255,0.35);
  background: rgba(4,5,15,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-size: 11px;
  padding: 6px 14px;
  line-height: 11px;
}

.emotion-2:hover {
  cursor: pointer;
}

<ViewMore
  showLessItemCount={1}
  showLessText="View less"
  showMoreText="View more"
>
  <Styled(ul)
    animate={true}
    blockHeight={0}
    innerRef={
      Object {
        "current": .emotion-0 {
  overflow: hidden;
  height: auto;
  -webkit-transition: height 400ms ease;
  transition: height 400ms ease;
}

<ul
          class="emotion-0 emotion-1"
        >
          <div>
            match
          </div>
          <div>
            match
          </div>
          <div>
            match
          </div>
          <div>
            match
          </div>
          <div>
            match
          </div>
          <div>
            match
          </div>
          <div>
            match
          </div>
          <div>
            match
          </div>
          <div>
            match
          </div>
          <div>
            match
          </div>
        </ul>,
      }
    }
    isExpanded={false}
  >
    <ul
      className="emotion-0 emotion-1"
    >
      <div
        key="0"
      >
        match
      </div>
      <div
        key="1"
      >
        match
      </div>
      <div
        key="2"
      >
        match
      </div>
      <div
        key="3"
      >
        match
      </div>
      <div
        key="4"
      >
        match
      </div>
      <div
        key="5"
      >
        match
      </div>
      <div
        key="6"
      >
        match
      </div>
      <div
        key="7"
      >
        match
      </div>
      <div
        key="8"
      >
        match
      </div>
      <div
        key="9"
      >
        match
      </div>
    </ul>
  </Styled(ul)>
  <Styled(div)>
    <div
      className="emotion-4 emotion-5"
    >
      <Styled(span)
        onClick={[Function]}
      >
        <span
          className="emotion-2 emotion-3"
          onClick={[Function]}
        >
          View more
        </span>
      </Styled(span)>
    </div>
  </Styled(div)>
</ViewMore>
`;
